tiatoolbox/tools/registration/wsi_registration.py

```diff
@@ -1233,3 +1248,184 @@ def apply_bspline_transform(
+        of the region.
+
+        Args:
+            location (tuple(int)):
+                (x, y) tuple giving the top left pixel in the baseline (level 0)
+                reference frame.
+            size (tuple(int)):
+                (width, height) tuple giving the desired output image size.
+            level (int):
+                Pyramid level/resolution layer.
+
+        Returns:
+            tuple:
+                - tuple(int) - transformed location (top left pixel).
+                - tuple(int) - maximum size suitable for transformation.
+
+        """
+        inv_transform = inv(self.transform_level0)
+        size_level0 = [x * (2**level) for x in size]
+        center_level0 = [x + size_level0[i] / 2 for i, x in enumerate(location)]
+        center_level0 = np.expand_dims(np.array(center_level0), axis=0)
+        center_level0 = self.transform_points(center_level0, inv_transform)[0]
+
+        transformed_size = self.get_patch_dimensions(size, inv_transform)
+        transformed_location = [
+            center_level0[0] - (transformed_size[0] * (2**level)) / 2,
+            center_level0[1] - (transformed_size[1] * (2**level)) / 2,
+        ]
+        transformed_location = tuple(
+            np.round(x).astype(int) for x in transformed_location
+        )
+        return transformed_location, transformed_size
+
+    def transform_patch(self, patch, size):
+        """Apply transformation to the given patch.
+
+        This function applies the transformation matrix after removing the translation.
+
+        Args:
+            patch (:class:`numpy.ndarray`):
+                A region of whole slide image.
+            size (tuple(int)):
+                (width, height) tuple giving the desired output image size.
+
+        Returns:
+            :class:`numpy.ndarray`:
+                A transformed region/patch.
+
+        """
+        transform = self.transform_level0 * [[1, 1, 0], [1, 1, 0], [1, 1, 1]]
+        forward_translation = np.array(
+            [[1, 0, -size[0] / 2], [0, 1, -size[1] / 2], [0, 0, 1]]
+        )
+        inverse_translation = np.array(
+            [[1, 0, size[0] / 2], [0, 1, size[1] / 2], [0, 0, 1]]
+        )
+        transform = inverse_translation @ transform @ forward_translation
+        return cv2.warpAffine(patch, transform[0:-1][:], patch.shape[:2][::-1])
+
+    def read_rect(self, location, size, level):
+        """Read a transformed region of the transformed whole slide image.
+
+        Location is in terms of the baseline image (level 0 / maximum resolution),
+        and size is the output image size.
+
+        Args:
+            location (tuple(int)):
+                (x, y) tuple giving the top left pixel in the baseline (level 0)
+                reference frame.
+            size (tuple(int)):
+                (width, height) tuple giving the desired output image size.
+            level (int):
+                Pyramid level/resolution layer.
```

Comment:
I think we should specify this in MPP rather than level or maybe use the units + resolution arguments as with `WSIReader` functions. What do you think @shaneahmed?

Reply:
yes I would prefer physical units as they are more consistent.

Reply:
https://github.com/TissueImageAnalytics/tiatoolbox/blob/e0106780f4aaeaf3bf6444ba1899c4ee1e346b89/tiatoolbox/tools/registration/wsi_registration.py#L1439-L1446